### ISSUE TITLE
Improve rust lib architecture

### ISSUE THREAD

**Issue by QuestofIranon** [bcmyers/argonautica#9]
Currently the `hasher` and `verifier` store both init configuration data which is likely never going to be mutated as well as mutable transaction information like the password being hashed. This causes an insane amount of cloning required to use it in a multi threaded context (the `to_owned()` functions).

An alternative would be to wrap the immutable configuration data inside of an internal structure within an [arc](https://doc.rust-lang.org/std/sync/struct.Arc.html). And treat data such as password and hash as a separate temporary transaction.

**Comment by WillSquire** [comment]
Yeah, I'm finding it harder to implement this than the other libraries purely because this takes less of a functional approach. I'd need to rewrite a lot of parent components just compensate. Otherwise I'd love to use this library.

**Comment by nbraud** [comment]
Yes; in general I would find it useful to have a configuration object that holds global (in the sense that it's reused across multiple hash computations) configuration value: secret key, and default parameters (parallelism `p`, tag length `τ`, memory size `m`, iterations count `t`, and type `y`, in [the spec]'s notation).

In particular, that would also make it easy to wipe the secret key when the configuration struct is dropped, probably by using the [`zeroize`] crate.

I'd be willing to implement this and submit a pull request.
@bcmyers Would you consider such a change?  Should I put the priority on the ergonomics and maintainability going forward, or should backwards compatibility be an overriding concern?

[the spec]: https://github.com/P-H-C/phc-winner-argon2/blob/master/argon2-specs.pdf
[`zeroize`]: https://docs.rs/zeroize/0.9.2/zeroize/

**Comment by bcmyers** [comment]
@nbraud - Would definitely welcome a contribution. This library needs some love and I haven't gotten around to providing it. Why don't you try for ergonomics and not worry too much about backwards compatibility. I feel like a major version update is the next step for this library, as there are a few rough edges I'd like to tackle with a clean slate. If anyone feels that's a bad idea, though, please let me know.
